jbi/retry.py

```diff
@@ -0,0 +1,84 @@
+import asyncio
+import logging
+from datetime import datetime, timedelta
+from os import getenv
+from time import sleep
+from typing import AsyncIterator
+
+import jbi.runner as runner
+from jbi.configuration import ACTIONS
+from jbi.queue import QueueItem, get_dl_queue
+
+CONSTANT_RETRY = getenv("DL_QUEUE_CONSTANT_RETRY", "false") == "true"
+RETRY_TIMEOUT_DAYS = getenv("DL_QUEUE_RETRY_TIMEOUT_DAYS", 7)
+
+
+async def retry_failed(item_executor=runner.execute_action):
+    queue = get_dl_queue()
+    logger = logging.getLogger(__name__)
+    min_event_timestamp = datetime.now() - timedelta(days=int(RETRY_TIMEOUT_DAYS))
+
+    # load all bugs from DLQ
+    bugs: dict[int, AsyncIterator[QueueItem]] = await queue.retrieve()
+
+    # metrics to track
+    metrics = {
+        "bug_count": len(bugs),
+        "events_processed": 0,
+        "events_skipped": 0,
+        "events_failed": 0,
+        "bugs_failed": 0,
+    }
+
+    for bug_id, items in bugs.items():
+        try:
+            async for item in items:
+                # skip and delete item if we have exceeded max_timeout
+                if item.timestamp < min_event_timestamp:
+                    logger.warning("removing expired event %s", item.identifier)
+                    await queue.done(item)
```

Comment:
Do we need another method for this? `discard()` or something with better semantics than `done()`? Especially if in the queue we log stuff like `item X is done`

Reply:
The queue should be just an IO operator and not have any real logic in it. It feels a little weird to have two different functions that do the same functional operation but with potentially different debug logs. Am I thinking about it wrong?

Looking at the logs the queue emits, we don't currently have a conflict problem. The debug level logs say things like "Removed {event} from queue for bug {bug_id}." and "Removed directory for bug {bug_id}". And our log levels will be above debug in prod and we won't alert on those.